plugins/ubiquiti/unifi_api

```diff
@@ -0,0 +1,1138 @@
+	$curl->setopt($curl->CURLOPT_SSL_VERIFYHOST, (( $APIconfig{"ssl_verify_host"} =~ m/no/i ) ? 0 : 2) );
+	$curl->setopt($curl->CURL_SSLVERSION_TLSv1, 1);
+	$curl->setopt($curl->CURLOPT_URL, $APIPoint{'login'});
+	$curl->setopt($curl->CURLOPT_POSTFIELDS, q[{'username':'] . $APIconfig{"user"} . q[', 'password':'] . $APIconfig{"pass"} . q['}] );
+	$curl->setopt($curl->CURLOPT_WRITEDATA, \$APIResponse{'login'});
+	$retcode = $curl->perform;
+
+	if ( $retcode != 0 ) {
+		die "FATAL:$me: Unable to connect to API: " . $curl->strerror($retcode) . " " . $curl->errbuf . "\n";
+	}
+
+	$APIJsonResponse{'login'} = $jsonOBJ->allow_nonref->utf8->relaxed->decode($APIResponse{'login'});
+
+	if ( $APIJsonResponse{'login'}->{'meta'}->{'rc'} ne 'ok' ) {
+		die "FATAL:$me: Unable to login to API - it said: " , $APIJsonResponse{'login'}->{'meta'}->{'msg'} , "\n";
+	}
+
+	# Change method to GET
+	$curl->setopt($curl->CURLOPT_HTTPGET,1);
+
+
+	# Get some API data.
+
+	# Device data
+	$curl->setopt($curl->CURLOPT_WRITEDATA, \$APIResponse{'device'});
+	$curl->setopt($curl->CURLOPT_URL, $APIPoint{'device'});
+	$retcode = $curl->perform;
+
+	if ( $retcode != 0 ) {
+		die "FATAL:$me: Unable to connect to API: " . $curl->strerror($retcode) . " " . $curl->errbuf . "\n";
+	}
+
+	$APIJsonResponse{'device'} = $jsonOBJ->allow_nonref->utf8->relaxed->decode($APIResponse{'device'});
+
+	if ( $APIJsonResponse{'device'}->{'meta'}->{'rc'} ne 'ok' ) {
+		die "FATAL:$me: Unable get device data from API - it said: " , $APIJsonResponse{'device'}->{'meta'}->{'msg'} , "\n";
+	}
+
+	# STA (client) data
+	$curl->setopt($curl->CURLOPT_WRITEDATA, \$APIResponse{'sta'});
+	$curl->setopt($curl->CURLOPT_URL, $APIPoint{'sta'});
+	$retcode = $curl->perform;
+
+	if ( $retcode != 0 ) {
+		die "FATAL:$me: Unable to connect to API: " . $curl->strerror($retcode) . " " . $curl->errbuf . "\n";
+	}
+
+	$APIJsonResponse{'sta'} = $jsonOBJ->allow_nonref->utf8->relaxed->decode($APIResponse{'sta'});
+
+	if ( $APIJsonResponse{'sta'}->{'meta'}->{'rc'} ne 'ok' ) {
+		die "FATAL:$me: Unable get sta data from API - it said: " , $APIJsonResponse{'sta'}->{'meta'}->{'msg'} , "\n";
+	}
+
+	# WLAN data
+	$curl->setopt($curl->CURLOPT_WRITEDATA, \$APIResponse{'wlan'});
+	$curl->setopt($curl->CURLOPT_URL, $APIPoint{'wlan'});
+	$retcode = $curl->perform;
+
+	if ( $retcode != 0 ) {
+		die "FATAL:$me: Unable to connect to API: " . $curl->strerror($retcode) . " " . $curl->errbuf . "\n";
+	}
+
+	$APIJsonResponse{'wlan'} = $jsonOBJ->allow_nonref->utf8->relaxed->decode($APIResponse{'wlan'});
+
+	if ( $APIJsonResponse{'wlan'}->{'meta'}->{'rc'} ne 'ok' ) {
+		die "FATAL:$me: Unable get wlan data from API - it said: " , $APIJsonResponse{'wlan'}->{'meta'}->{'msg'} , "\n";
+	}
+}
+
+# Make field names safe, and lowercase. Sourced originally from Munin::Plugin
+sub make_safe { $_ = lc($_[0]); s/^[^a-z_]+/_/; s/\W/_/g; return ( $_ eq "root" ? "__root" : $_ ); }
```

Comment:
Did you decide against `clean_fieldname`?

Reply:
Yes, and actually, this isn't the finished version the more I think about this - I was normalizing to all lowercase with this - but the unifi controller software has no stipulation that the names I am using for field names are unique.  So, I'm going to dig a little deeper into what the API is spitting out to find out if there is a uniq suffix I can add to them (I don't want to drop the device name, as it makes finding the RRD in question a *lot* easier) - maybe the MAC address as well

Reply:
Maybe the `need_multigraph` hint from @niclan helps you to embrace the beauty of reusing existing code? :)

Regarding unique names: it is surely a good idea to use prefixes in situations were independent names may try to squat the same namespace.

Reply:
Agreed.  Testing through what I just did, but I did indeed use clean_fieldname - this function is still there, because I've added a way to toggle off what I consider correct behavior and use the "old" naming method - otherwise, I'm going to need to hand rename a couple hundred RRD files.